### PR TITLE
[deconz] Support for air quality sensor

### LINKED ISSUE #11882: [deconz] Support for air quality sensor (ZHAAirQuality -> airquality and airqualityppb)
## Environment

* Version used: openHAB 3.2.0
* Binding: Dresden Elektronik deCONZ Binding 
* Hardware: ConBee II v2.13.04
* Exampel sensor: air quality sensor (AQSZB-110) from Develco Products A/S
* Since [ConBee II v2.6.0](https://github.com/dresden-elektronik/deconz-rest-plugin/pull/3607) the airquality sensor is supported. 

## Request

Support for reading the type _ZHAAirQuality_ and the states _airquality_ as label and _airqualityppb_ as number.

Via the REST API of ConBee II the data would have to be read out and provided. 

```json
{
    "config": {
        "battery": 100,
        "on": true,
        "reachable": true
    },
    "ep": 38,
    "etag": "46799cb65bc....",
    "lastannounced": "2021-12-28T23:59:02Z",
    "lastseen": "2021-12-29T12:04Z",
    "manufacturername": "Develco Products A/S",
    "modelid": "AQSZB-110",
    "name": "AirQuality 4",
    "state": {
        "airquality": "good",
        "airqualityppb": 178,
        "lastupdated": "2021-12-29T12:00:45.462"
    },
    "swversion": "2021-10-28 08:59",
    "type": "ZHAAirQuality",
    "uniqueid": "00:00:00:00:00:00:00:00-26-fc03"
}
```

### Doc Upgrade
In addition, the doc would have to be updated for this. 

## Result
grafik

### PR DESCRIPTION
**Pull request by nixoso**
[deconz] Support for air quality sensor implemented

Resolves #11882

The enhancement allows as described in #11882, the reading of air quality sensors via e.g. ConBee II.
On the one hand the air quality is determined as string label, as well as the current air quality as number in ppb. 

 The documentation has been expanded accordingly to include the air quality sensor and the new read-out options. 

Added two new unit tests to check the reading of the:
* airquality label
* airqualityppb number

### CODE REVIEW

**Comment by jlaur** [comment]
@nixoso - can you add "Fixes #11882" or "Resolves #11882" in description to have the issue linked correctly?

**Comment by nixoso** [comment]
@jlaur 
Is done.

File: bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml [hunk 1]
```diff
@@ -491,6 +491,41 @@
...
+		<description>Current air quality level</description>
+		<category>Airquality</category>
+		<state readOnly="true" pattern="%s"></state>
+	</channel-type>
+
+	<channel-type id="airqualityppb">
+		<item-type>Number</item-type>
+		<label>Air quality in ppb</label>
+		<description>Current air quality ppb (parts per billion)</description>
+		<category>Airquality</category>
```
Comment (jlaur) [inline bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml:524]: Which kind of particles are measured? If CO2 consider category **carbondioxide** - I don't think category Airquality exists?

File: bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml [hunk 2]
```diff
@@ -491,6 +491,41 @@
...
+	<channel-type id="airquality">
+		<item-type>String</item-type>
+		<label>Air quality level</label>
+		<description>Current air quality level</description>
+		<category>Airquality</category>
+		<state readOnly="true" pattern="%s"></state>
+	</channel-type>
+
+	<channel-type id="airqualityppb">
+		<item-type>Number</item-type>
```
Comment (jlaur) [inline bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml:521]: Number:Dimensionless?

File: bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml (hunk 1 above)
Comment (nixoso) [reply bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml]: I have removed the category, at the moment there is no matching one for it. Te air quality based on measurements of volatile organic compounds (VOCs). The measured value is specified in ppb (parts per billion).

File: bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml (hunk 2 above)
Comment (nixoso) [reply bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml]: Item-type was changed to Number:Dimensionless.

File: bundles/org.openhab.binding.deconz/README.md [hunk 3]
```diff
@@ -218,6 +221,7 @@ Bridge deconz:deconz:homeserver [ host="192.168.0.10", apikey="ABCDEFGHIJ" ] {
     presencesensor      livingroom-presence     "Livingroom Presence"       [ id="1" ]
     temperaturesensor   livingroom-temperature  "Livingroom Temperature"    [ id="2" ]
     humiditysensor      livingroom-humidity     "Livingroom Humidity"       [ id="3" ]
+    airqualitysensor    livingroom-voc          "Livingroom Voc"            [ id="4" ]
     pressuresensor      livingroom-pressure     "Livingroom Pressure"       [ id="4" ]
```
Comment (jlaur) [inline bundles/org.openhab.binding.deconz/README.md:225]: Do you need to renumber id's as there are now two with `id="4"`?

Comment (nixoso) [reply bundles/org.openhab.binding.deconz/README.md]: I have assigned a new unique id for the example.

File: bundles/org.openhab.binding.deconz/README.md [hunk 4]
```diff
@@ -235,6 +239,8 @@ Bridge deconz:deconz:homeserver [ host="192.168.0.10", apikey="ABCDEFGHIJ" ] {
 Switch                  Livingroom_Presence     "Presence Livingroom [%s]"          <motion>        { channel="deconz:presencesensor:homeserver:livingroom-presence:presence" }
 Number:Temperature      Livingroom_Temperature  "Temperature Livingroom [%.1f °C]"  <temperature>   { channel="deconz:temperaturesensor:homeserver:livingroom-temperature:temperature" }
 Number:Dimensionless    Livingroom_Humidity     "Humidity Livingroom [%.1f %%]"     <humidity>      { channel="deconz:humiditysensor:homeserver:livingroom-humidity:humidity" }
+String                  LIVING_ROOM_VOC_LABEL    "Air quality Livingroom [%s]"      <none>          { channel="deconz:airqualitysensor:homeserver:livingroom-voc:airquality" }
+Number                  LIVING_ROOM_VOC          "Air quality [%d ppb]"             <none>          {           channel="deconz:airqualitysensor:homeserver:livingroom-voc:airqualityppb" }
```
Comment (jlaur) [inline bundles/org.openhab.binding.deconz/README.md:243]: Other example items are prefixed "Livingroom", but here casing is "LIVING_ROOM". Suggest to harmonize.

Comment (jlaur) [inline bundles/org.openhab.binding.deconz/README.md:243]: This is now `Number:Dimensionless`.

Comment (jlaur) [inline bundles/org.openhab.binding.deconz/README.md:243]: This is now `Number:Dimensionless`.

File: bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/discovery/ThingDiscoveryService.java [hunk 5]
```diff
@@ -281,6 +281,8 @@ private void addSensor(String sensorID, SensorMessage sensor) {
             thingTypeUID = THING_TYPE_BATTERY_SENSOR; // ZHABattery
         } else if (sensor.type.contains("ZHAThermostat")) {
             thingTypeUID = THING_TYPE_THERMOSTAT; // ZHAThermostat
+        } else if (sensor.type.contains("ZHAAirQuality")) {
+            thingTypeUID = THING_TYPE_AIRQUALITY_SENSOR; // ZHAAirQuality
```
Comment (jlaur) [inline bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/discovery/ThingDiscoveryService.java:285]: Comment seems redundant?

File: bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml [hunk 6]
```diff
@@ -491,6 +491,41 @@
 		<state readOnly="true"/>
 	</channel-type>
 
+
+	<thing-type id="airqualitysensor">
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="deconz"/>
+		</supported-bridge-type-refs>
+		<label>Air quality Sensor</label>
+		<description>A air quality sensor</description>
```
Comment (jlaur) [inline bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml:500]: Grammar: A**n** air quality sensor.

Comment (jlaur) [inline bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml:500]: Texts must be updated in **src/main/resources/OH-INF/i18n/deconz.properties** also.

File: bundles/org.openhab.binding.deconz/README.md (hunk 4 above)
Comment (nixoso) [reply bundles/org.openhab.binding.deconz/README.md]: Good point I haromnised the doc.

Comment (nixoso) [reply bundles/org.openhab.binding.deconz/README.md]: Doc is adapted.

File: bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/discovery/ThingDiscoveryService.java (hunk 5 above)
Comment (nixoso) [reply bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/discovery/ThingDiscoveryService.java]: Comment is removed.

File: bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml (hunk 6 above)
Comment (nixoso) [reply bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml]: Fixed

Comment (nixoso) [reply bundles/org.openhab.binding.deconz/src/main/resources/OH-INF/thing/sensor-thing-types.xml]: Translations are done.

File: bundles/org.openhab.binding.deconz/README.md (hunk 4 above)
Comment (jlaur) [reply bundles/org.openhab.binding.deconz/README.md]: Seems like it was not pushed?

Comment (nixoso) [reply bundles/org.openhab.binding.deconz/README.md]: With Commit https://github.com/openhab/openhab-addons/pull/11885/commits/75b1cb7075e39b2178c2e76f8a2b2364a78fff18 it was pushed.

Comment (jlaur) [reply bundles/org.openhab.binding.deconz/README.md]: Hmm, my comment was for **airqualityppb** being documented as `Number`, when it's now `Number:Dimensionless`. I don't see this change in current version or in that commit.

File: bundles/org.openhab.binding.deconz/README.md [hunk 7]
```diff
@@ -150,6 +151,8 @@ The sensor devices support some of the following channels:
 | battery_level   | Number                   | R           | Battery level (in %)                                                                      | any battery-powered sensor                        |
 | battery_low     | Switch                   | R           | Battery level low: `ON`; `OFF`                                                            | any battery-powered sensor                        |
 | carbonmonoxide  | Switch                   | R           | `ON` = carbon monoxide detected                                                           | carbonmonoxide                                    |
+| airquality      | String                   | R           | Current air quality level                                                                 | airqualitysensor                                  |
+| airqualityppb   | Integer                  | R           | Current air quality ppb (parts per billion)                                               | airqualitysensor                                  |
```
Comment (jlaur) [inline bundles/org.openhab.binding.deconz/README.md:155]:
```suggestion
| airqualityppb   | Number:Dimensionless     | R           | Current air quality ppb (parts per billion)                                               | airqualitysensor                                  |
```

Comment (nixoso) [reply bundles/org.openhab.binding.deconz/README.md]: Is fixed

File: bundles/org.openhab.binding.deconz/README.md (hunk 4 above)
Comment (nixoso) [reply bundles/org.openhab.binding.deconz/README.md]: Now I understood it and changed it.

File: bundles/org.openhab.binding.deconz/README.md [hunk 8]
```diff
@@ -235,6 +239,8 @@ Bridge deconz:deconz:homeserver [ host="192.168.0.10", apikey="ABCDEFGHIJ" ] {
 Switch                  Livingroom_Presence     "Presence Livingroom [%s]"          <motion>        { channel="deconz:presencesensor:homeserver:livingroom-presence:presence" }
 Number:Temperature      Livingroom_Temperature  "Temperature Livingroom [%.1f °C]"  <temperature>   { channel="deconz:temperaturesensor:homeserver:livingroom-temperature:temperature" }
 Number:Dimensionless    Livingroom_Humidity     "Humidity Livingroom [%.1f %%]"     <humidity>      { channel="deconz:humiditysensor:homeserver:livingroom-humidity:humidity" }
+String                  Livingroom_voc_label    "Air quality Livingroom [%s]"       <none>          { channel="deconz:airqualitysensor:homeserver:livingroom-voc:airquality" }
+Number:Dimensionless    Livingroom_voc          "Air quality [%d ppb]"              <none>          {           channel="deconz:airqualitysensor:homeserver:livingroom-voc:airqualityppb" }
```
Comment (jlaur) [inline bundles/org.openhab.binding.deconz/README.md:243]: Sorry to bother you again, but now when reading the adapted and formatted README two new comments:
- I think you should remove "<none>", i.e. no category.
- There seem to be some alignment problem with the last line: multiple spaces before "{ channel=".

Comment (nixoso) [reply bundles/org.openhab.binding.deconz/README.md]: I fixed the indentation in the layout.
And I removed the None category from the example.

**Review by jlaur — APPROVED** [review]
Looks good to me, but I'm not able to test it.

**Comment by nixoso** [comment]
> Looks good to me, but I'm not able to test it.

Thanks for the review. In the screenshot in the feature request (https://github.com/openhab/openhab-addons/issues/11882#issue-1090481541) you can see the result of the implementation.

**Comment by jlaur** [comment]
> Thanks for the review. In the screenshot in the feature request ([#11882 (comment)](https://github.com/openhab/openhab-addons/issues/11882#issue-1090481541)) you can see the result of the implementation.

Thanks, it looks very good. I would love to merge your PR, but actually I just became a maintainer yesterday, i.e. after adding all my review comments. So before merging I need to learn a few things from other maintainers, most importantly how to correctly edit the final commit message, in this case whether to preserve each message as-is (with the "openhab#11882 - " prefix) or whether there is freedom to edit it more freely and for example remove those prefixes. It might seem like a small thing, but I would really like to get it right before exercising my new privileges as maintainer. Hang in tight. :-)

**Pull request merged**